machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,389 @@
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Represents the latest available observations of current state.
+	// +optional
+	Conditions []ControllerConfigStatusCondition `json:"conditions,omitempty"`
+}
+
+// ControllerConfigStatusCondition contains condition information for ControllerConfigStatus
+type ControllerConfigStatusCondition struct {
+	// type specifies the state of the operator's reconciliation functionality.
+	Type ControllerConfigStatusConditionType `json:"type"`
+
+	// status of the condition, one of True, False, Unknown.
+	Status corev1.ConditionStatus `json:"status"`
+
+	// lastTransitionTime is the time of the last update to the current status object.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	// reason is the reason for the condition's last transition.  Reasons are CamelCase
+	Reason string `json:"reason,omitempty"`
+
+	// message provides additional information about the current condition.
+	// This is only to be consumed by humans.
+	Message string `json:"message,omitempty"`
+}
+
+// ControllerConfigStatusConditionType valid conditions of a machineconfigpool
+type ControllerConfigStatusConditionType string
+
+const (
+	TemplateContollerRunning   ControllerConfigStatusConditionType = "TemplateContollerRunning"
+	TemplateContollerCompleted ControllerConfigStatusConditionType = "TemplateContollerCompleted"
+	TemplateContollerFailing   ControllerConfigStatusConditionType = "TemplateContollerFailing"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ControllerConfigList is a list of ControllerConfig resources
+type ControllerConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []ControllerConfig `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigPool describes a pool of MachineConfigs.
+type MachineConfigPool struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MachineConfigPoolSpec   `json:"spec"`
+	Status MachineConfigPoolStatus `json:"status"`
+}
+
+// MachineConfigPoolSpec is the spec for MachineConfigPool resource.
+type MachineConfigPoolSpec struct {
+	// Label selector for MachineConfigs.
+	// Refer https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/ on how label and selectors work.
+	MachineConfigSelector *metav1.LabelSelector `json:"machineConfigSelector,omitempty"`
+
+	// Node selector for Machines
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+	// If true, changes to this machine config pool should be stopped.
+	// This includes generating new desiredMachineConfig and update of machines.
+	Paused bool `json:"paused"`
+
+	// MaxUnavailable specifies the percentage or constant number of machines that can be updating at any given time.
```

Comment:
the comment should use the json name of the field because they show up in `kubectl explain` and docs.